Fix excessive RAM usage and slow processing in `sct_process_segmentation` by resampling after cropping

sandrinebedard opened the pull request:
## Checklist

<!-- Hi, and thank you for submitting a Pull Request! Please make sure to check the boxes below before submitting. -->

- [x] **PR Sidebar**: I've filled these options within the PR sidebar:
  - [Labels](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#233-labels) (**exactly one** dark purple label!)
  - [Milestone](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#234-milestone)
- [x] **Guidelines**: I've read the [Contributing guidelines](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)) to make sure my [branch](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing%3A-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#part-1-branches) and [pull request](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing%3A-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#part-2-pull-requests) meet SCT's guidelines. (Feel free to stop and fixup your commits before submitting.)
- [x] **Documentation**: I've updated the [relevant documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages.
- [x] **Tests**: I've added [relevant tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Testing) for my contribution, if necessary. I've also made sure that my contribution passes [all of the automated tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#24-checking-the-automated-tests) (which will be triggered after the PR is submitted).
- [ ] Only **after** the automated test suite passes: I will tag a [reviewer](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Contributing:-Submitting-changes-to-SCT-(Opening-a-Pull-Request)#231-reviewers) in the PR sidebar. (You can also ask for help with specific questions in the PR comments before the tests pass.)

## Description

This PR moves resampling to perform it on cropped 2D patches  instead of on original segmentation, which is much faster.

### TODO
* update values in PAM50 normalized metrics with new AP and symmetry metrics: https://github.com/spinalcordtoolbox/PAM50-normalized-metrics/issues/29


## Linked issues

Fixes #5120

sandrinebedard committed:
change for resampling after cropping

sandrinebedard committed:
fix

sandrinebedard committed:
remove extra ,

sandrinebedard commented:
@jcohenadad @joshuacwnewton @valosekj Looks like the resampling was the problem!

When testing on the T2 from the course data, it went from 56 s to 2.3 s to run the command!

@joshuacwnewton however, with the method I am using for resampling, the CSA values differ quite a lot, I was wondering if you had a better solution, or if I was missing something! (see the tests)

mguaypaq commented on spinalcordtoolbox/process_seg.py:
```diff
@@ -300,7 +300,18 @@ def _properties2d(seg, dim, iz, angle_hog=None, verbose=1):
-    seg_crop_r = seg_crop
+    # Apply resampling to the cropped segmentation:
+    zoom_factors = (dim[0]/0.1, dim[1]/0.1)
+    seg_crop_r = zoom(seg_crop, zoom=zoom_factors, order=1)  # make pixel size isotropic
```
I think the large change is due to this:
```suggestion
    seg_crop_r = zoom(seg_crop, zoom=zoom_factors, order=1, mode='grid-constant', grid_mode=True)  # make pixel size isotropic
```

According to [the `zoom` docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.zoom.html#zoom), the default `grid_mode=False` essentially chops off one half-voxel from each edge of the data array when computing distances, and this has a large effect because the cropped segmentation can be quite small.

In the case of `sct_example_data/t2/t2_seg.nii.gz`:
- a cropped slice has shape (16, 13)
- after resampling, this is increased to (128, 104)
- with `grid_mode=False`, the zoom factors applied are 127/15 = 8.467 in the x-axis and 103/12 = 8.583 in the y-axis, which is wrong
- but with `grid_mode=True`, the zoom factors are 128/16 = 8.000 and 104/13 = 8.000, which is correct.

I think the default for all `scipy.ndimage` resampling functions is to assume that array values are pinpoint-like and correspond to a corner of the pixel/voxel (this is what `mode={constant,mirror,wrap}` all do). But for nifti images, the array values correspond to the center of the voxel, and are considered to take up the whole voxel (which is that `mode={grid-constant,grid-mirror,grid-wrap}` all do).

(This is why I recently used `mode='grid-constant'` [in the QC code](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/9fff8f26b9b1f427bd7afe156e38c7845639e186/spinalcordtoolbox/reports/qc2.py#L473-L475).)

I tried out this change for the `t2/csa.csv[MEAN(area)]` values, and got:
- 60.4292865764842 on master
- 67.13273546108884 on this PR without `grid_mode`
- 60.50923891297235 on this PR with `grid_mode=True`.

mguaypaq reviewed (commented):
This is what I found regarding the large change in mean CSA values with this PR.

sandrinebedard committed:
Update spinalcordtoolbox/process_seg.py

fix grid_mode

Co-authored-by: Mathieu Guay-Paquet <mathieu.guay-paquet@polymtl.ca>

sandrinebedard committed:
updae test values

sandrinebedard committed:
clean up

sandrinebedard committed:
use pr variable

sandrinebedard committed:
fix arg

sandrinebedard committed:
remove resampling of image as only used for HOG, cropping doesn't make sense for it

sandrinebedard committed:
fix values

sandrinebedard committed:
change resmaplign as now only performed on crop

sandrinebedard committed:
update value

sandrinebedard committed:
update values

sandrinebedard committed:
fix values

joshuacwnewton commented on testing/cli/test_cli_sct_process_segmentation.py:
Here is the "dummy segmentation":

<img width="975" height="493" alt="Image" src="https://github.com/user-attachments/assets/b9d2b096-1657-47f3-98ad-e47e4863cfb9" />

With default radii of `(3.0, 5.0)` the final rectangular shape is `(7.0, 11.0)` hence the "base" CSA value of `77`. So that makes sense! The angle value are also exactly as expected, too.

Given an `angle_AP` of 15*, `cos(15)` = .965, meaning the area should shrink to about 96.5% of 77 vox, which gives ~74 vox. Meaning, I think both the old and the new values make roughly the same amount of sense?

tl;dr I don't think we should be using a "base" value of 77 when we know what the expected angle correction should be less than this. This should be an easy fix I think.

joshuacwnewton replied:
This touches the same code involved in @mguaypaq's investigation in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/5129#discussion_r2706010351.

While `mode='grid-constant'` does keep the current values in line with the `master` values, is there a chance that the `master` values were incorrect to begin with? I feel like maybe there's something I don't fully understand yet... shouldn't resampling largely preserve the area of the original smaller array?

mguaypaq replied:
The only immediate use of an `scipy.ndimage` function I could find which might need a different `mode` parameter (though I'm not sure) is this:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/25485f40f6255013fafdf41cc5ba44b184c5184a/spinalcordtoolbox/process_seg.py#L425-L428

Given that it's not in the code touched by this PR, maybe we could just merge this and continue the investigation in a new issue.

joshuacwnewton reviewed (commented):
I've identified the code responsible for the failing test case, but I don't yet understand whether this is expected behavior or unexpected behavior. (Hopefully @mguaypaq can help?)

Either way, I don't think this PR is to blame:

> - **Expected**: `73.15 <= x <= 80.85`
> - **master**: `73.19`
> - **This PR**: `73.11`

So, I _think_ this is either a larger problem with the tests or with the code itself, and was present on `master` before this PR?

joshuacwnewton commented on spinalcordtoolbox/process_seg.py:
```diff
@@ -300,7 +293,17 @@ def _properties2d(seg, dim, iz, angle_hog=None, verbose=1):
+    minx, miny, maxx, maxy = region.bbox
+    # Use those bounding box coordinates to crop the segmentation (for faster processing), again as zoom adds padding
+    seg_crop_r = seg_crop_r[np.clip(minx-pad, 0, seg_crop_r.shape[0]): np.clip(maxx+pad, 0, seg_crop_r.shape[0]),
+                            np.clip(miny-pad, 0, seg_crop_r.shape[1]): np.clip(maxy+pad, 0, seg_crop_r.shape[1])]
```
Uh-oh. I'm seeing the same wildly varying angle values on our tutorial data: https://www.diffchecker.com/79Gom7Bh/

joshuacwnewton replied:
> I'll push a test to this branch (aiming for Monday), and then we can apply the fixes and make sure everything still looks good (or perhaps better!) vs. 7.1.

This ended up being a lot more involved than I anticipated, but it's on this branch now! Now we can try to update the code to get this test to pass. :)

joshuacwnewton committed:
Amend PR #5129 to properly test for issues with segmentation metrics (#5180)

This PR tries to update #5129 to make sure our test properly catches
metric issues by:

- Decreasing the relative tolerance we use to check values
- Checking per-value metrics instead of mean values when possible
- Doing slice-by-slice comparisons to ensure metrics don't vary.
- Removing the hardcoded ground-truth values to avoid normalization of
deviance when the tests fail.

This test fully passes on v7.1 and the tolerance could probably go a bit
lower, but fails on `master` and for PR #5129. I think my ideal would be
to reduce the tolerance further, then make the code changes necessary
for things to pass, but we should be confident that we're testing the
correct things before merging the test.

## Linked issues

N/A.
To be merged into #5129.

---------

Co-authored-by: Mathieu Guay-Paquet <mathieu.guay-paquet@polymtl.ca>

sandrinebedard committed:
remove 2nd cropping

sandrinebedard committed:
change thr for binarization for bbox

sandrinebedard commented:
@joshuacwnewton @mguaypaq I have now removed the 2nd cropping, and change the binarization threshold to get the bounding box for cropping. But it doesn't look like the test is passing 😢 , did I miss something, or this was the change we wanted right?

mguaypaq commented on testing/cli/test_cli_sct_compute_compression.py:
I know it wasn't touched by this PR, but I just want to point out that these two lines are weird: the second one exactly overwrites the first.

sandrinebedard replied:
probably a mistake

mguaypaq replied:
I think this will be fixed when we update the tests to use `-discfile` instead of `-vertfile` (following the deprecation in #4948), so I'm gonna say this is out of scope for this PR.

sandrinebedard commented on spinalcordtoolbox/process_seg.py:
but now the AP is not the one used from `region.minor_axis_length`, we directly sum the voxels instead (changes introduced in 7.2):
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/1a608486b84a46e25bf8e452f699bc2af67b13fe/spinalcordtoolbox/process_seg.py#L344-L349

mguaypaq replied:
Ok, thanks to your comment I was able to figure out what's happening. TLDR: we should update the dummy segmentation used by the tests, because it's too symmetric. I will add commits to do this.

- The tests for `sct_compute_compression` look at the AP diameter through a metric ratio (in other words, "how many percentage points is the compressed slice less than the surrounding slices"):

  https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/2db6a78be189940517fc9f89329622b32c7875cc/spinalcordtoolbox/scripts/sct_compute_compression.py#L516-L525

- The dummy segmentation consists of a vertical column with a rectangular cross-section.
  - For most slices, the cross-section is a 15x15 square (voxels, at a 1mm resolution).
  - The single "compressed" slice is a 12x15 rectangle (with 2 rows of voxels shaved off in the A direction, and 1 row of voxels shaved off in the P direction).

- On `master`:
  - The regular slices have a detected angle of 90 degrees (so, the AP diameter is actually computed in the LR direction). The resulting length is 14.91, which is close to the true value of 15 (but, accidentally).
  - The compressed slice correctly has a detected angle of 0 degrees. The resulting length is 11.96, which is close to the true value of 12.
  - The computed percentage is $1 - 11.96/14.91 = 19.79\\%$, which is close to the expected value of $1 - 12/15 = 20\\%$.

- On this branch:
  - The regular slices have a detected angle of 45 degrees (so, the AP diameter is actually computed along the diagonal of the square). The resulting length is 19.6. You might expect a length of $15\cdot\sqrt{2} = 21.2$, but you have to subtract 1.5 to account for the averaging of the AP measurement across 3mm. Anyway, the test expects 15, which is very far from this.
  - The compressed slice is correctly close to 12.
  - The computed percentage is $1 - 12/19.6 = 38.78\\%$, very far from the expected value.

- The real problem is that the square slices are so symmetric that the detected angle can easily be 0, 45, 90, etc. degrees, and this is sensitive to small changes in resampling.

- I suggest changing the dummy spinal cord mask to remove one layer of voxels on each of the A and P sides, so that it's wider in the LR direction than the AP direction, as expected by the angle detection code.
  - When I tried doing this, the `process_seg` code reliably detected 0 degrees of rotation on all the slices, and the `sct_compute_compression` code correctly detected a ratio of $1 - 10/13 = 23.08\\%$.

mguaypaq replied:
With my added commits, I've updated the test and the expected results for AP diameter (with a more reasonable tolerance).

There is still a failure, for `area_ratio`:

```
___ test_sct_compute_compression ___
Traceback (most recent call last):
  ...
AssertionError: assert 23.076883697405737 == 20.05342163028501 ± 2.0e-05
  
  comparison failed
  Obtained: 23.076883697405737
  Expected: 20.05342163028501 ± 2.0e-05
```

(I'll investigate this difference next week.)

mguaypaq replied:
Ok, this was actually very easy:
- The failure is in `area_ratio`.
- When I modified the dummy segmentation to make it rectangle, instead of square, I changed the area of the "compressed" slice and the surrounding slices, so this is expected.
- The old expected value corresponds to the old area ratio:
  - Compressed slice area is $12 \times 15 = 180$.
  - Surrounding slice area is $15 \times 15 = 225$.
  - Expected ratio is then $1 - 180/225 = 20.00\\%$.
- The new obtained value fits with the new dummy CSA values:
  - Compressed slice area is $10 \times 15 = 150$.
  - Surrounding slice are is $13 \times 15 = 195$.
  - Expected ratio is then $1 - 150/195 = 23.08\\%$.
- I'll push a commit to update the values and tolerance for this test.

mguaypaq replied:
Ok, the tests should now pass with the updated values and tolerances. After that, I'm happy to merge this PR. Sorry for the wait!

mguaypaq committed:
Modify dummy SC mask for `sct_compute_compression` tests

The old one was too symmetric (square cross-section), so the
detected orientation of each slice was unreliable (0, 45, 90 deg).

The new one is slightly thinner in the AP direction, so that the
detected orientation is reliably 0 degrees.

mguaypaq committed:
Update test values and tolerances

mguaypaq committed:
Update test values

mguaypaq reviewed (approved):
I'm reasonably happy with the updated tests.

mguaypaq committed:
Document the provenance of test values

mguaypaq committed:
Remove uint8 test

We have no hope of getting a robust expected value for this test.

mguaypaq committed:
Update more test values

I was using an old version of `PAM50_normalized_metrics` before.

mguaypaq committed:
Merge branch 'master' into sb/5120-resampling-process-seg

mguaypaq committed:
Update batch_processing results
